src/ds/rb_map.h

```diff
@@ -284,6 +303,353 @@ namespace rb
+          return newMap.fixDoubleBlackLeft();
+        }
+        // no double blacks are present
+        return std::make_pair(newMap, false);
+      }
+      else if (rootk < key)
+      {
+        // mirror of the above case
+        auto right_without = right()._remove(key);
+        auto newMap =
+          Map(rootColor(), left(), rootKey(), rootValue(), right_without.first);
+        if (right_without.second)
+        {
+          return newMap.fixDoubleBlackRight();
+        }
+        return std::make_pair(newMap, false);
+      }
+      else if (key == rootk)
+      {
+        // delete key from this node
+        if (left().empty() && right().empty())
+        {
+          // leaf node, a simple case
+          auto doubleBlack = rootColor() == B;
+          return std::make_pair(Map(), doubleBlack);
+        }
+        else if (left().empty())
+        {
+          // nothing on the left so we can replace this node with the right
+          // child
+          auto r = right();
+          // Exactly one of the node being removed and the right node are black:
+          // - the left is empty so has a black height of 1
+          // - the right must also have a black height of 1 to maintain the
+          // height but it is not empty or we would have hit the above if
+          // statement
+          // - therefore the right node is red.
+          assert(r.left().empty());
+          assert(r.right().empty());
+          assert(r.rootColor() == R);
+          return std::make_pair(r.paint(B), false);
+        }
+        else if (right().empty())
+        {
+          // mirror of the above case
+          return std::make_pair(left().paint(B), false);
+        }
+        else
+        {
+          // both children are non-empty, swap this node's key and value with
+          // the successor and then delete the successor
+          auto successor = right().minimum();
+          auto right_without = right()._remove(successor.first);
+          auto newMap = Map(
+            rootColor(),
+            left(),
+            successor.first,
+            successor.second,
+            right_without.first);
+          if (right_without.second)
+          {
+            return newMap.fixDoubleBlackRight();
+          }
+          return std::make_pair(newMap, false);
+        }
+      }
+      else
+      {
+        // key not found in the tree
+        return std::make_pair(Map(*this), false);
+      }
+    }
+
+    // Return the minimum key in this map along with its value.
+    std::pair<K, V> minimum() const
+    {
+      assert(!empty());
+
+      if (left().empty())
+      {
+        return std::make_pair(rootKey(), rootValue());
+      }
+      else
+      {
+        return left().minimum();
+      }
+    }
+
+    // Check properties of the tree
+    void check() const
+    {
+      int totalBlackCount;
+      _check(0, totalBlackCount);
+    }
+
+    // Print an s-expression style representation of the tree's colors
```

Comment:
Perhaps only available for debug builds?

Reply:
What's the preferred way to make it only available for debug builds?

Reply:
```c++
#ifndef NDEBUG 
    ... debug code
#else
    ... release code
#endif
```